server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java

```diff
@@ -442,9 +444,20 @@ public ClusterState addIndexTemplateV2(final ClusterState currentState, final bo
 
         validate(name, finalIndexTemplate);
         logger.info("adding index template [{}]", name);
-        return ClusterState.builder(currentState)
+        ClusterState newState = ClusterState.builder(currentState)
             .metadata(Metadata.builder(currentState.metadata()).put(name, finalIndexTemplate))
             .build();
+        if (finalIndexTemplate.getDataStreamTemplate() != null) {
+            validateDataStreamTemplate(name, finalIndexTemplate, newState);
+        }
```

Comment:
This validation needs to be moved to the `validateCompositeTemplate` method, that way, when a component template is updated we will still validate the mapping is correct.

Otherwise a composable template could be added where the mapping is in the component template, and then the component template could be updated to remove the timestamp mapping. Once we move the validation it'll be checked for both index and component templates

Reply:
Thanks for catching this. I've placed the validation in the wrong place.